apps/condo/domains/common/components/PostMessageProvider/Provider.tsx

```diff
@@ -0,0 +1,225 @@
+    event: Method,
+    origin: '*' | string,
+    handler: RequestHandler<Method>
+) => void
+
+/**
+ * Context definitions
+ */
+type IPostMessageContext = {
+    allowedOrigins: Readonly<Array<string>>
+    registerOrigin: (origin: string) => void
+    resetOrigin: (origin: string) => void
+    handlers: Readonly<Record<HandlerOrigin, OriginHandlers>>
+    registerHandler: RegisterHandler
+    validators: Readonly<ValidatorsType>
+}
+
+/**
+ * Mocks in case Provider was not provided by DOM
+ */
+const PostMessageContext = createContext<IPostMessageContext>({
+    allowedOrigins: [serverUrl],
+    registerOrigin: () => ({}),
+    resetOrigin: () => ({}),
+    handlers: {},
+    registerHandler: () => ({}),
+    validators,
+})
+/**
+ * Common incoming message pattern and its schema
+ */
+
+type MessageType = {
+    handler: string
+    params: RequestId & Record<string, unknown>
+    type: 'condo-bridge' | 'condo-ui'
+    version: string
+}
+
+const ajv = new Ajv()
+const messageSchema = {
+    type: 'object',
+    properties: {
+        handler: { type: 'string' },
+        params: { type: 'object' },
+        type: { type: 'string', enum: ['condo-bridge', 'condo-ui'] },
+        version: { type: 'string', pattern: '^\\d+.\\d+.\\d+(?:-\\w+){0,1}$' },
+    },
+    required: ['handler', 'type', 'version', 'params'],
+    additionalProperties: false,
+}
+const validateMessage: ValidateFunction<MessageType> = ajv.compile(messageSchema)
+
+function getClientErrorMessage<Method extends AllRequestMethods> (
+    reason: ErrorReason,
+    message: string,
+    requestId?: RequestIdType,
+    method?: Method
+): ClientErrorResponse<AllRequestMethods, ErrorReason> {
+    return {
+        type: method ? `${method}Error` : 'CondoWebAppCommonError',
+        data: {
+            errorType: 'client',
+            errorCode: ERROR_CODES[reason],
+            errorReason: reason,
+            errorMessage: message,
+            ...(typeof requestId !== 'undefined' ? { requestId } : null),
+        },
+    }
+}
+
+/**
+ * Single place to handle all incoming postMessages from condo and its miniapps.
+ * You can register global handler for all origins by specifying "*" as origin (Example: notification)
+ * Or pass "per origin" handler (Example: window resize)
+ * Provider takes care of 3 things:
+ * 1. Making sure messages are trusted and coming from allowed Windows
+ * 2. Making sure message can be handled and passing parameters are valid
+ * 3. Handling errors thrown by individual handlers and converting it to corresponding postMessage response
+ */
+export const PostMessageProvider: React.FC = ({ children }) => {
+    const [allowedOrigins, setAllowedOrigins] = useState<Array<string>>([serverUrl])
+    const [registeredHandlers, setRegisteredHandlers] = useState<Record<HandlerOrigin, OriginHandlers>>({})
+    const isOnClient = typeof window !== 'undefined'
+
+    const registerHandler: RegisterHandler = useCallback((event, origin, handler) => {
```

Comment:
What about to make naming consistent with DOM, like `add…`/`…remove` instead of `register…`/`reset…`?
For example: `addEventHandler`, `removeEventHandler`, `addOrigin`, `removeOrigin`.